lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -410,31 +410,72 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
   // For each node:
   for (auto &N : nodes) {
     auto *node = &N;
-    // Sink Transpose below batch normalization nodes:
+    // Sink Reshape/Transpose below BatchNormalization.
     if (auto *BN = dyn_cast<BatchNormalizationNode>(node)) {
-      auto *TR = dyn_cast<TransposeNode>(BN->getInput());
 
-      if (!TR) {
+      // Sink Reshape below BatchNormalization.
+      if (auto *RS = dyn_cast<ReshapeNode>(BN->getInput())) {
+        auto inDims = RS->getInput().dims();
+        auto outDims = RS->getResult().dims();
+
+        // Reshape should not change the BatchNorm ChannelIdx dimensions.
+
+        // Check for NH[W]C layout.
+        if (outDims[BN->getChannelIdx()] == outDims.size() - 1) {
+          if (inDims.size() - 1 != outDims.size() - 1) {
+            continue;
+          }
+        }
+
+        // Check for NCH[W] layout.
+        if (outDims[BN->getChannelIdx()] == outDims[1]) {
+          if (inDims[1] != outDims[1]) {
+            continue;
+          }
+        }
```

Comment:
Same -- shouldn't this be something like
```suggestion
        // Note: 1 here is for C in NCH[W]
        if (BN->getChannelIdx() != 1 && inDims[1] == outDims[1]) {
          continue;
        }
```

Reply:
Assuming Conv -> Reshape -> BatchNorm sequence, for NCH[W], if we allow sink of Reshape below BatchNorm then, BatchNorm will fold into Conv which should not happen. BatchNorm folding into Convolution is allowed only when the layout is NHWC. For now I am not allowing the optimization to kick in for NCHW case.

Reply:
>BatchNorm folding into Convolution is allowed only when the layout is NHWC.

@ugaresuyash Maybe I'm missing something, but why should that be the case? NHWC vs NCHW is just about the layout of the data in memory, why would the fusion of BN require NHWC?

Assuming it is the case for some reason, it seems a bit scary to rely on there being a reshape there to prevent the conv-batch norm fusion from kicking in. Seems to me like we should allow this reshape sinking to be done regardless of the layout, but then also update the conv-batch norm fusion to only kick in when we have an NHWC.

Reply:
@jfix, BatchNorm fold assumes that the layout has to be NHWC when fusing BatchNorm parameters into Convolution weights.  Please refer to 'normalizeWeights()'.

What you say also makes sense. I will look into adding the sink optimization and then prevent Conv + BatchNorm fusion.

Reply:
Oh, you just meant what glow currently supports, got it.

Reply:
@jfix71 ,
I did attempt to allow sink of reshape below batchnorm for NCHW case. Sink works well. However, during normalizeWeights(), I see the channel idx of BN as 3 instead of 1 which was set in the test body. Something strange is happening. Either there is some other issue, potentially, which is not preserving the BN channel idx. Or may be I am missing something.

For now, for my internal work, I think NHWC serves the purpose. Can we land NHWC for now and then later create an issue to handle the NCHW case separately?

Reply:
@jfix71, I have made the necessary changes for adding NCHW case and a corresponding test. 
Please take look at it.